File: cross_platform_build/lib/views/character_creator_view.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:google_fonts/google_fonts.dart';
import 'package:provider/provider.dart';
import '../managers/user_profile_manager.dart';
import '../models/lote_models.dart';
import 'widgets/pixel_sprite_widget.dart';

class CharacterCreatorView extends StatefulWidget {
  const CharacterCreatorView({super.key});

  @override
  State<CharacterCreatorView> createState() => _CharacterCreatorViewState();
}

class _CharacterCreatorViewState extends State<CharacterCreatorView> {
  int _selectedToolColor = 1; // Palette index: 1=Skin, 2=Hair, 3=Eyes, 4=Outfit, 5=Aura, 0=Eraser
  late List<List<int>> _pixelGrid;

  // Preset values
  String _sexPreset = "Male";
  String _planetPreset = "Ninjonia";
  String _skinPreset = "Fair";
  String _hairStylePreset = "Spiky";
  String _hairColorPreset = "Yellow";
  String _outfitColorPreset = "Gray";

  @override
  void initState() {
    super.initState();
    final profile = Provider.of<UserProfileManager>(context, listen: false);
    // Deep copy pixel grid
    _pixelGrid = List.generate(profile.sprite.pixelGrid.length, (r) => List.from(profile.sprite.pixelGrid[r]));

    // Try to infer presets from loaded hex values
    _sexPreset = profile.sprite.sex;
    _planetPreset = profile.sprite.outfitStyle.isEmpty ? profile.homePlanet : profile.sprite.outfitStyle;
    _skinPreset = _skinPresetFromHex(profile.sprite.skinColorHex);
    _hairStylePreset = profile.sprite.hairStyle;
    _hairColorPreset = _hairColorFromHex(profile.sprite.hairColorHex);
    _outfitColorPreset = _outfitColorFromHex(profile.sprite.outfitColorHex);
  }

  // Preset hex mappings
  String _skinHex(String style) {
    switch (style) {
      case "Tan":
        return "#D7CCC8";
      case "Dark":
        return "#8D6E63";
      case "Celestial":
        return "#9575CD";
      case "Golden":
        return "#FFE082";
      case "Pale":
        return "#FFECB3";
      default:
        return "#FFD180"; // Fair
    }
  }

  String _skinPresetFromHex(String hex) {
    if (hex == "#D7CCC8") return "Tan";
    if (hex == "#8D6E63") return "Dark";
    if (hex == "#9575CD") return "Celestial";
    if (hex == "#FFE082") return "Golden";
    if (hex == "#FFECB3") return "Pale";
    return "Fair";
  }

  String _hairHex(String color) {
    switch (color) {
      case "Red":
        return "#FF3D00";
      case "Black":
        return "#212121";
      case "Blue":
        return "#29B6F6";
      case "Silver":
        return "#CFD8DC";
      case "Green":
        return "#66BB6A";
      case "Brown":
        return "#5D4037";
      case "Orange":
        return "#FF9100";
      default:
        return "#FFEA00"; // Yellow
    }
  }

  String _hairColorFromHex(String hex) {
    if (hex == "#FF3D00") return "Red";
    if (hex == "#212121") return "Black";
    if (hex == "#29B6F6") return "Blue";
    if (hex == "#CFD8DC") return "Silver";
    if (hex == "#66BB6A") return "Green";
    if (hex == "#5D4037") return "Brown";
    if (hex == "#FF9100") return "Orange";
    return "Yellow";
  }

  String _outfitColorHex(String style) {
    switch (style) {
      case "Dark Blue":
        return "#1A237E";
      case "Crimson":
        return "#B71C1C";
      case "Gold":
        return "#FFD700";
      case "Purple":
        return "#4A148C";
      case "Green":
        return "#1B5E20";
      case "Charcoal":
        return "#263238";
      case "Steel":
        return "#455A64";
      default:
        return "#37474F"; // Gray
    }
  }

  String _outfitColorFromHex(String hex) {
    if (hex == "#1A237E") return "Dark Blue";
    if (hex == "#B71C1C") return "Crimson";
    if (hex == "#FFD700") return "Gold";
    if (hex == "#4A148C") return "Purple";
    if (hex == "#1B5E20") return "Green";
    if (hex == "#263238") return "Charcoal";
    if (hex == "#455A64") return "Steel";
    return "Gray";
  }

  Color _activeSkinColor() => hexToColor(_skinHex(_skinPreset));
  Color _activeHairColor() => hexToColor(_hairHex(_hairColorPreset));
  Color _activeOutfitColor() => hexToColor(_outfitColorHex(_outfitColorPreset));

  Color _colorForPixelValue(int val, UserProfileManager profile) {
    switch (val) {
      case 1:
        return _activeSkinColor();
      case 2:
        return _activeHairColor();
      case 3:
        return profile.currentElement.primaryColor;
      case 4:
        return _activeOutfitColor();
      case 5:
        return profile.currentElement.accentColor;
      default:
        return Colors.white.withOpacity(0.08);
    }
  }

  void _paintPixel(int r, int c) {
    setState(() {
      _pixelGrid[r][c] = _selectedToolColor;
    });
  }

  void _handlePaintGesture(Offset localPosition, double boxSize) {
    final int gridSize = _pixelGrid.length;
    if (gridSize == 0) return;
    final double pixelWidth = boxSize / gridSize;
    
    final int col = (localPosition.dx / pixelWidth).floor();
    final int row = (localPosition.dy / pixelWidth).floor();
    
    if (row >= 0 && row < gridSize && col >= 0 && col < gridSize) {
      if (_pixelGrid[row][col] != _selectedToolColor) {
        setState(() {
          _pixelGrid[row][col] = _selectedToolColor;
        });
      }
    }
  }

  void _clearCanvas() {
    setState(() {
      _pixelGrid = List.generate(300, (_) => List.generate(300, (_) => 0));
    });
  }

  void _generateBaseFromPresets() {
    setState(() {
      _pixelGrid = CharacterSprite.generateGrid(
        planet: _planetPreset,
        hairStyle: _hairStylePreset,
        sex: _sexPreset,
      );
    });
  }

  void _saveSprite(UserProfileManager profile) {
    profile.sprite.pixelGrid = _pixelGrid;
    profile.sprite.skinColorHex = _skinHex(_skinPreset);
    profile.sprite.hairColorHex = _hairHex(_hairColorPreset);
    profile.sprite.outfitColorHex = _outfitColorHex(_outfitColorPreset);
    profile.sprite.hairStyle = _hairStylePreset;
    profile.sprite.outfitStyle = _planetPreset;
    profile.sprite.sex = _sexPreset;
    profile.sprite = profile.sprite; // Trigger didSet save/notify
    Navigator.of(context).pop();
  }

  @override
  Widget build(BuildContext context) {
    final profile = Provider.of<UserProfileManager>(context);
    final activeSkinColor = hexToColor(_skinHex(_skinPreset));
    final activeHairColor = hexToColor(_hairHex(_hairColorPreset));
    final activeOutfitColor = hexToColor(_outfitColorHex(_outfitColorPreset));
    final activeAuraColor = profile.currentElement.accentColor;
    final activeEyeColor = profile.currentElement.primaryColor;

    return Scaffold(
      backgroundColor: const Color(0xFF050505),
      appBar: AppBar(
        backgroundColor: const Color(0xFF050505),
        elevation: 0,
        leading: IconButton(
          icon: const Icon(Icons.close, color: Colors.white),
          onPressed: () => Navigator.of(context).pop(),
        ),
      ),
      body: SafeArea(
        child: SingleChildScrollView(
          padding: const EdgeInsets.symmetric(horizontal: 20, vertical: 10),
          child: Column(
            children: [
              // Header
              Text(
                "CHARACTER FORGE",
                style: GoogleFonts.orbitron(
                  fontSize: 24,
                  fontWeight: FontWeight.bold,
                  letterSpacing: 4,
                  color: Colors.white,
                ),
              ),
              const SizedBox(height: 6),
              Text(
                "Design your Elsaither Warrior's pixel avatar",
                style: GoogleFonts.exo2(
                  fontSize: 13,
                  color: Colors.grey,
                ),
              ),
              const SizedBox(height: 25),

              // Canvas drawing grid
              Center(
                child: Container(
                  padding: const EdgeInsets.all(10),
                  decoration: BoxDecoration(
                    color: Colors.black.withOpacity(0.6),
                    borderRadius: BorderRadius.circular(8),
                    border: Border.all(
                      color: profile.currentElement.primaryColor.withOpacity(0.4),
                      width: 1.5,
                    ),
                  ),
                  child: GestureDetector(
                    onPanDown: (details) {
                      _handlePaintGesture(details.localPosition, 300.0);
                    },
                    onPanUpdate: (details) {
                      _handlePaintGesture(details.localPosition, 300.0);
                    },
                    child: PixelSpriteWidget(
                      grid: _pixelGrid,
                      skinColor: activeSkinColor,
                      hairColor: activeHairColor,
                      eyeColor: activeEyeColor,
                      outfitColor: activeOutfitColor,
                      auraColor: activeAuraColor,
                      pixelSize: 1.0,
                    ),
                  ),
                ),
              ),
              const SizedBox(height: 25),

              // Brush Palettes
              Align(
                alignment: Alignment.centerLeft,
                child: Text(
                  "PALETTE BRUSHES",
                  style: GoogleFonts.orbitron(
                    fontSize: 12,
                    fontWeight: FontWeight.bold,
                    color: Colors.grey,
                    letterSpacing: 2,
                  ),
                ),
              ),
              const SizedBox(height: 10),
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  _buildBrushButton(1, "Skin", _activeSkinColor()),
                  _buildBrushButton(2, "Hair", _activeHairColor()),
                  _buildBrushButton(3, "Eyes", profile.currentElement.primaryColor),
                  _buildBrushButton(4, "Armor", _activeOutfitColor()),
                  _buildBrushButton(5, "Aura", profile.currentElement.accentColor),
                  _buildBrushButton(0, "Eraser", Colors.black, isEraser: true),
                ],
              ),
              const SizedBox(height: 25),

              // Presets Card
              Container(
                padding: const EdgeInsets.all(16),
                decoration: BoxDecoration(
                  color: Colors.white.withOpacity(0.02),
                  borderRadius: BorderRadius.circular(12),
                  border: Border.all(color: Colors.white.withOpacity(0.05)),
                ),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text(
                      "PRESETS & LAYERS",
                      style: GoogleFonts.orbitron(
                        fontSize: 14,
                        fontWeight: FontWeight.bold,
                        color: Colors.white,
                        letterSpacing: 2,
                      ),
                    ),
                    _buildPresetDropdown("Sex", _sexPreset, ["Male", "Female"], (val) {
                      setState(() {
                        _sexPreset = val!;
                      });
                    }, profile),
                    _buildPresetDropdown("Planet Style", _planetPreset, ["Ninjonia", "Techno", "Warrion", "Battacaria"], (val) {
                      setState(() {
                        _planetPreset = val!;
                      });
                    }, profile),
                    _buildPresetDropdown("Skin Tone", _skinPreset, ["Fair", "Tan", "Dark", "Celestial", "Golden", "Pale"], (val) {
                      setState(() {
                        _skinPreset = val!;
                      });
                    }, profile),
                    _buildPresetDropdown("Hair Style", _hairStylePreset, ["Spiky", "Long", "Short", "Mohawk"], (val) {
                      setState(() {
                        _hairStylePreset = val!;
                      });
                    }, profile),
                    _buildPresetDropdown("Hair Color", _hairColorPreset, ["Yellow", "Red", "Black", "Blue", "Silver", "Green", "Brown", "Orange"], (val) {
                      setState(() {
                        _hairColorPreset = val!;
                      });
                    }, profile),
                    _buildPresetDropdown("Clothing Color", _outfitColorPreset, ["Gray", "Dark Blue", "Crimson", "Gold", "Purple", "Green", "Charcoal", "Steel"], (val) {
                      setState(() {
                        _outfitColorPreset = val!;
                      });
                    }, profile),
                    const SizedBox(height: 16),
                    Row(
                      children: [
                        Expanded(
                          child: ElevatedButton(
                            onPressed: _generateBaseFromPresets,
                            style: ElevatedButton.styleFrom(
                              backgroundColor: profile.currentElement.primaryColor.withOpacity(0.2),
                              side: BorderSide(color: profile.currentElement.primaryColor),
                              shape: RoundedRectangleBorder(
                                borderRadius: BorderRadius.circular(8),
                              ),
                            ),
                            child: Text(
                              "COMPOSITE PRESETS",
                              style: GoogleFonts.orbitron(
                                fontSize: 10,
                                fontWeight: FontWeight.bold,
                                color: Colors.white,
                              ),
                            ),
                          ),
                        ),
                        const SizedBox(width: 12),
                        Expanded(
                          child: OutlinedButton(
                            onPressed: _clearCanvas,
                            style: OutlinedButton.styleFrom(
                              side: BorderSide(color: Colors.grey.withOpacity(0.5)),
                              shape: RoundedRectangleBorder(
                                borderRadius: BorderRadius.circular(8),
                              ),
                            ),
                            child: Text(
                              "CLEAR CANVAS",
                              style: GoogleFonts.orbitron(
                                fontSize: 10,
                                fontWeight: FontWeight.bold,
                                color: Colors.grey,
                              ),
                            ),
                          ),
                        )
                      ],
                    )
                  ],
                ),
              ),
              const SizedBox(height: 30),

              // Save button
              SizedBox(
                width: double.infinity,
                child: ElevatedButton(
                  onPressed: () => _saveSprite(profile),
                  style: ElevatedButton.styleFrom(
                    backgroundColor: profile.currentElement.primaryColor,
                    padding: const EdgeInsets.symmetric(vertical: 14),
                    shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(10),
                    ),
                    elevation: 6,
                    shadowColor: profile.currentElement.primaryColor.withOpacity(0.4),
                  ),
                  child: Text(
                    "FORGE ARTIFACT",
                    style: GoogleFonts.orbitron(
                      fontSize: 15,
                      fontWeight: FontWeight.bold,
                      letterSpacing: 2,
                      color: Colors.white,
                    ),
                  ),
                ),
              ),
              const SizedBox(height: 20),
            ],
          ),
        ),
      ),
    );
  }

  Widget _buildBrushButton(int val, String label, Color color, {bool isEraser = false}) {
    final isSelected = _selectedToolColor == val;
    return GestureDetector(
      onTap: () {
        setState(() {
          _selectedToolColor = val;
        });
      },
      child: Column(
        children: [
          Container(
            width: 36,
            height: 36,
            decoration: BoxDecoration(
              color: isEraser ? Colors.white.withOpacity(0.15) : color,
              borderRadius: BorderRadius.circular(6),
              border: isSelected
                  ? Border.all(color: Colors.white, width: 2)
                  : Border.all(color: Colors.transparent, width: 2),
            ),
            child: isEraser
                ? const Icon(Icons.cleaning_services_rounded, color: Colors.white, size: 18)
                : null,
          ),
          const SizedBox(height: 4),
          Text(
            label,
            style: TextStyle(
              fontSize: 10,
              color: isSelected ? Colors.white : Colors.grey,
            ),
          )
        ],
      ),
    );
  }

  Widget _buildPresetDropdown(
    String title,
    String currentValue,
    List<String> options,
    ValueChanged<String?> onChanged,
    UserProfileManager profile,
  ) {
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 6),
      child: Container(
        padding: const EdgeInsets.symmetric(horizontal: 12),
        decoration: BoxDecoration(
          color: Colors.white.withOpacity(0.03),
          borderRadius: BorderRadius.circular(8),
        ),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Text(
              title,
              style: GoogleFonts.exo2(fontSize: 13, color: Colors.white),
            ),
            DropdownButton<String>(
              value: currentValue,
              items: options.map((opt) {
                return DropdownMenuItem<String>(
                  value: opt,
                  child: Text(
                    opt,
                    style: const TextStyle(color: Colors.white),
                  ),
                );
              }).toList(),
              onChanged: onChanged,
              dropdownColor: const Color(0xFF0F0F0F),
              underline: const SizedBox(),
              iconEnabledColor: profile.currentElement.primaryColor,
            )
          ],
        ),
      ),
    );
  }
}

File: deprecated_swift_version/LotE Workout Companion/Views/CharacterCreatorView.swift
//
//  CharacterCreatorView.swift
//  LotE Workout Companion
//
//  Created by Antigravity on 6/23/26.
//

import SwiftUI

struct CharacterCreatorView: View {
    @ObservedObject var profileManager: UserProfileManager
    @Environment(\.presentationMode) var presentationMode
    
    @State private var selectedToolColor: Int = 1 // Palette index: 1=Skin, 2=Hair, 3=Eyes, 4=Outfit, 5=Aura, 0=Eraser
    @State private var pixelGrid: [[Int]] = CharacterSprite.defaultGrid
    
    // Preset configuration state
    @State private var sexPreset: String = "Male"
    @State private var planetPreset: String = "Ninjonia"
    @State private var skinPreset: String = "Fair"
    @State private var hairStylePreset: String = "Spiky"
    @State private var hairColorPreset: String = "Yellow"
    @State private var outfitColorPreset: String = "Gray"
    
    // Colors mapped from selections
    var activeSkinColor: Color { Color(hex: skinHex(for: skinPreset)) ?? .orange }
    var activeHairColor: Color { Color(hex: hairHex(for: hairColorPreset)) ?? .yellow }
    var activeOutfitColor: Color { Color(hex: outfitHex(for: outfitColorPreset)) ?? .gray }
    var activeAuraColor: Color { profileManager.currentElement.accentColor }
    var activeEyeColor: Color { profileManager.currentElement.primaryColor }
    
    var body: some View {
        ZStack {
            Color(hex: "#050505")
                .ignoresSafeArea()
            
            ScrollView {
                VStack(spacing: 20) {
                    // Header
                    VStack(spacing: 8) {
                        Text("CHARACTER FORGE")
                            .font(.custom("Orbitron-Bold", size: 24).bold())
                            .foregroundColor(.white)
                            .tracking(4)
                        
                        Text("Design your Elsaither Warrior's pixel avatar")
                            .font(.custom("Exo2-Medium", size: 13))
                            .foregroundColor(.gray)
                    }
                    .padding(.top, 25)
                    
                    // Sprite Preview & Pixel Grid
                    VStack(spacing: 15) {
                        // High performance character canvas
                        let pixelSize: CGFloat = 0.8
                        let size: CGFloat = 300.0 * pixelSize
                        Canvas { context, sz in
                            for r in 0..<300 {
                                for c in 0..<300 {
                                    let val = pixelGrid[r][c]
                                    if val != 0 {
                                        let color: Color = {
                                            switch val {
                                            case 1: return activeSkinColor
                                            case 2: return activeHairColor
                                            case 3: return activeEyeColor
                                            case 4: return activeOutfitColor
                                            case 5: return activeAuraColor
                                            default: return .clear
                                            }
                                        }()
                                        let rect = CGRect(
                                            x: CGFloat(c) * pixelSize,
                                            y: CGFloat(r) * pixelSize,
                                            width: pixelSize,
                                            height: pixelSize
                                        )
                                        context.fill(Path(rect), with: .color(color))
                                    }
                                }
                            }
                        }
                        .frame(width: size, height: size)
                        .gesture(
                            DragGesture(minimumDistance: 0)
                                .onChanged { value in
                                    let location = value.location
                                    let col = Int(location.x / pixelSize)
                                    let row = Int(location.y / pixelSize)
                                    if row >= 0 && row < 300 && col >= 0 && col < 300 {
                                        if pixelGrid[row][col] != selectedToolColor {
                                            pixelGrid[row][col] = selectedToolColor
                                        }
                                    }
                                }
                        )
                        .padding(10)
                        .background(Color.black.opacity(0.6))
                        .cornerRadius(8)
                        .overlay(
                            RoundedRectangle(cornerRadius: 8)
                                .stroke(profileManager.currentElement.primaryColor.opacity(0.4), lineWidth: 1.5)
                        )
                    }
                    
                    // Color Tool Selectors
                    VStack(alignment: .leading, spacing: 8) {
                        Text("PALETTE BRUSHES")
                            .font(.custom("Orbitron-Bold", size: 12).bold())
                            .foregroundColor(.gray)
                            .tracking(2)
                            .padding(.horizontal)
                        
                        HStack(spacing: 12) {
                            brushButton(val: 1, label: "Skin", color: activeSkinColor)
                            brushButton(val: 2, label: "Hair", color: activeHairColor)
                            brushButton(val: 3, label: "Eyes", color: activeEyeColor)
                            brushButton(val: 4, label: "Armor", color: activeOutfitColor)
                            brushButton(val: 5, label: "Aura", color: activeAuraColor)
                            brushButton(val: 0, label: "Eraser", color: .black, isEraser: true)
                        }
                        .padding(.horizontal)
                    }
                    
                    // Layer Presets Accordion
                    VStack(spacing: 15) {
                        Text("PRESETS & LAYERS")
                            .font(.custom("Orbitron-Bold", size: 14).bold())
                            .foregroundColor(.white)
                            .tracking(2)
                            .frame(maxWidth: .infinity, alignment: .leading)
                        
                        // Preset selectors
                        VStack(spacing: 12) {
                            presetSelector(title: "Sex", selection: $sexPreset, options: ["Male", "Female"])
                            presetSelector(title: "Planet Style", selection: $planetPreset, options: ["Ninjonia", "Techno", "Warrion", "Battacaria"])
                            presetSelector(title: "Skin Tone", selection: $skinPreset, options: ["Fair", "Tan", "Dark", "Celestial", "Golden", "Pale"])
                            presetSelector(title: "Hair Style", selection: $hairStylePreset, options: ["Spiky", "Long", "Short", "Mohawk"])
                            presetSelector(title: "Hair Color", selection: $hairColorPreset, options: ["Yellow", "Red", "Black", "Blue", "Silver", "Green", "Brown", "Orange"])
                            presetSelector(title: "Clothing Color", selection: $outfitColorPreset, options: ["Gray", "Dark Blue", "Crimson", "Gold", "Purple", "Green", "Charcoal", "Steel"])
                        }
                        
                        // Reset Grid or Re-generate preset base
                        HStack(spacing: 15) {
                            Button(action: {
                                generateBaseFromPresets()
                            }) {
                                Text("COMPOSITE PRESETS")
                                    .font(.custom("Orbitron-Bold", size: 11).bold())
                                    .foregroundColor(.white)
                                    .padding(.vertical, 10)
                                    .frame(maxWidth: .infinity)
                                    .background(
                                        RoundedRectangle(cornerRadius: 8)
                                            .fill(profileManager.currentElement.primaryColor.opacity(0.3))
                                            .overlay(
                                                RoundedRectangle(cornerRadius: 8)
                                                    .stroke(profileManager.currentElement.primaryColor, lineWidth: 1)
                                            )
                                    )
                            }
                            
                            Button(action: {
                                clearCanvas()
                            }) {
                                Text("CLEAR CANVAS")
                                    .font(.custom("Orbitron-Bold", size: 11).bold())
                                    .foregroundColor(.gray)
                                    .padding(.vertical, 10)
                                    .frame(maxWidth: .infinity)
                                    .background(
                                        RoundedRectangle(cornerRadius: 8)
                                            .stroke(Color.gray.opacity(0.5), lineWidth: 1)
                                    )
                            }
                        }
                        .padding(.top, 5)
                    }
                    .padding(20)
                    .background(Color.white.opacity(0.02))
                    .cornerRadius(12)
                    .padding(.horizontal)
                    
                    // Save Button
                    Button(action: {
                        saveSprite()
                    }) {
                        Text("FORGE ARTIFACT")
                            .font(.custom("Orbitron-Bold", size: 15).bold())
                            .foregroundColor(.white)
                            .tracking(2)
                            .padding(.vertical, 14)
                            .frame(maxWidth: .infinity)
                            .background(
                                RoundedRectangle(cornerRadius: 10)
                                    .fill(profileManager.currentElement.primaryColor)
                                    .shadow(color: profileManager.currentElement.primaryColor.opacity(0.4), radius: 6)
                            )
                    }
                    .padding(.horizontal, 20)
                    .padding(.bottom, 30)
                }
            }
        }
        .onAppear {
            pixelGrid = profileManager.sprite.pixelGrid
            sexPreset = profileManager.sprite.sex
            planetPreset = profileManager.sprite.outfitStyle.isEmpty ? profileManager.homePlanet : profileManager.sprite.outfitStyle
            skinPreset = skinPresetFromHex(profileManager.sprite.skinColorHex)
            hairStylePreset = profileManager.sprite.hairStyle
            hairColorPreset = hairColorPresetFromHex(profileManager.sprite.hairColorHex)
            outfitColorPreset = outfitColorPresetFromHex(profileManager.sprite.outfitColorHex)
        }
    }
    
    // MARK: - Brush Button Helper
    @ViewBuilder
    private func brushButton(val: Int, label: String, color: Color, isEraser: Bool = false) -> some View {
        Button(action: {
            selectedToolColor = val
        }) {
            VStack(spacing: 5) {
                if isEraser {
                    Image(systemName: "eraser.line.dashed")
                        .frame(width: 32, height: 32)
                        .background(Color.white.opacity(0.15))
                        .cornerRadius(6)
                        .foregroundColor(.white)
                } else {
                    RoundedRectangle(cornerRadius: 6)
                        .fill(color)
                        .frame(width: 32, height: 32)
                        .overlay(
                            RoundedRectangle(cornerRadius: 6)
                                .stroke(Color.white.opacity(0.5), lineWidth: selectedToolColor == val ? 2 : 0)
                        )
                }
                
                Text(label)
                    .font(.system(size: 10))
                    .foregroundColor(selectedToolColor == val ? .white : .gray)
            }
        }
    }
    
    // MARK: - Preset Dropdown Selector
    private func presetSelector(title: String, selection: Binding<String>, options: [String]) -> some View {
        HStack {
            Text(title)
                .font(.custom("Exo2-Medium", size: 14))
                .foregroundColor(.white)
            Spacer()
            Picker(title, selection: selection) {
                ForEach(options, id: \.self) { opt in
                    Text(opt).tag(opt)
                }
            }
            .pickerStyle(MenuPickerStyle())
            .accentColor(profileManager.currentElement.primaryColor)
        }
        .padding(.vertical, 8)
        .padding(.horizontal)
        .background(Color.white.opacity(0.03))
        .cornerRadius(8)
    }
    
    // MARK: - Pixel Coloring Helpers
    private func colorForPixelValue(_ val: Int) -> Color {
        switch val {
        case 1: return activeSkinColor
        case 2: return activeHairColor
        case 3: return activeEyeColor
        case 4: return activeOutfitColor
        case 5: return activeAuraColor
        default: return Color.white.opacity(0.08)
        }
    }
    
    private func paintPixel(row: Int, col: Int) {
        if pixelGrid[row][col] != selectedToolColor {
            pixelGrid[row][col] = selectedToolColor
        }
    }
    
    private func clearCanvas() {
        pixelGrid = Array(repeating: Array(repeating: 0, count: 300), count: 300)
    }
    
    // MARK: - Preset Compositor
    private func generateBaseFromPresets() {
        pixelGrid = CharacterSprite.generateGrid(planet: planetPreset, hairStyle: hairStylePreset, sex: sexPreset)
    }
    
    // Hex mappings for presets
    private func skinHex(for style: String) -> String {
        switch style {
        case "Tan": return "#D7CCC8"
        case "Dark": return "#8D6E63"
        case "Celestial": return "#9575CD"
        case "Golden": return "#FFE082"
        case "Pale": return "#FFECB3"
        default: return "#FFD180" // Fair
        }
    }
    
    private func skinPresetFromHex(_ hex: String) -> String {
        switch hex {
        case "#D7CCC8": return "Tan"
        case "#8D6E63": return "Dark"
        case "#9575CD": return "Celestial"
        case "#FFE082": return "Golden"
        case "#FFECB3": return "Pale"
        default: return "Fair"
        }
    }
    
    private func hairHex(for color: String) -> String {
        switch color {
        case "Red": return "#FF3D00"
        case "Black": return "#212121"
        case "Blue": return "#29B6F6"
        case "Silver": return "#CFD8DC"
        case "Green": return "#66BB6A"
        case "Brown": return "#5D4037"
        case "Orange": return "#FF9100"
        default: return "#FFEA00" // Yellow
        }
    }
    
    private func hairColorPresetFromHex(_ hex: String) -> String {
        switch hex {
        case "#FF3D00": return "Red"
        case "#212121": return "Black"
        case "#29B6F6": return "Blue"
        case "#CFD8DC": return "Silver"
        case "#66BB6A": return "Green"
        case "#5D4037": return "Brown"
        case "#FF9100": return "Orange"
        default: return "Yellow"
        }
    }
    
    private func outfitHex(for color: String) -> String {
        switch color {
        case "Dark Blue": return "#1A237E"
        case "Crimson": return "#B71C1C"
        case "Gold": return "#FFD700"
        case "Purple": return "#4A148C"
        case "Green": return "#1B5E20"
        case "Charcoal": return "#263238"
        case "Steel": return "#455A64"
        default: return "#37474F" // Gray
        }
    }
    
    private func outfitColorPresetFromHex(_ hex: String) -> String {
        switch hex {
        case "#1A237E": return "Dark Blue"
        case "#B71C1C": return "Crimson"
        case "#FFD700": return "Gold"
        case "#4A148C": return "Purple"
        case "#1B5E20": return "Green"
        case "#263238": return "Charcoal"
        case "#455A64": return "Steel"
        default: return "Gray"
        }
    }
    
    // MARK: - Save Sprite Action
    private func saveSprite() {
        profileManager.sprite.pixelGrid = pixelGrid
        profileManager.sprite.skinColorHex = skinHex(for: skinPreset)
        profileManager.sprite.hairColorHex = hairHex(for: hairColorPreset)
        profileManager.sprite.outfitColorHex = outfitHex(for: outfitColorPreset)
        profileManager.sprite.hairStyle = hairStylePreset
        profileManager.sprite.outfitStyle = planetPreset
        profileManager.sprite.sex = sexPreset
        presentationMode.wrappedValue.dismiss()
    }
}
